subprojects/core/src/main/java/org/gradle/api/internal/file/archive/ZipFileTree.java

```diff
@@ -46,19 +47,22 @@ public class ZipFileTree extends AbstractArchiveFileTree {
     private final Chmod chmod;
     private final DirectoryFileTreeFactory directoryFileTreeFactory;
     private final FileHasher fileHasher;
+    private final TemporaryFileProvider temporaryExtractionDir;
 
     public ZipFileTree(
         Provider<File> zipFile,
         Chmod chmod,
         DirectoryFileTreeFactory directoryFileTreeFactory,
         FileHasher fileHasher,
-        DecompressionCache decompressionCache
+        DecompressionCache decompressionCache,
+        TemporaryFileProvider temporaryExtractionDir
```

Comment:
🤔 It seems a little weird to me that the `XYZFileTree` objects possess both a `DecompressionCache` and a `TemporaryFileProvider`.  The `TemporaryFileProvider` seems like it should belong to the cache.

Would it be worth reworking this to so that the trees are only responsible for caculating the `"zip_" + fileHasher.hash(zipFile);` part of the path, and the `temporaryExtractionDir.newTemporaryDirectory(".cache", "expanded"` part is done internally by the cache itself, which then just recieves the file-specific part and concatenates them together?

Reply:
Yeah, I went down this path at some point but there are a few problems:
1. The DecompressionCache is build-tree scoped, so it doesn't know the full path to the expanded directory.
2. If we wanted to pass that along (through `useCache(...)`) then we're back to the same problem of needing to provide a `TemporaryFileProvider` to pass later. 

We could instead have a project-scoped `DecompressionCache` service which knows the temporary file path for the project and can build the file path, but we need the expandedDir in XXXFileTree and it gets a little weird to query `DecompressionCache` with one half of the information.

If we were doing this from scratch, I don't think we'd make the caches per-project, so I don't think it makes sense to make the DecompressionCache more project specific. 

As I was going through this again, I thought about renaming `DecompressionCache` to `DecompressionCoordinator` or `DecompressionManager` or something like that. That removes the idea that the `DecompressionCache` is responsible for any of the content.  It's only handling the coordination about when it's OK to extract for a particular key (directory).

Reply:
> I thought about renaming DecompressionCache to DecompressionCoordinator or DecompressionManager or something like that

I really like this idea.

Reply:
done